File: source/ml/factory/factory.cpp
/** 
 * @brief Machine learning factory implementation details.
 */
#include <memory>

#include "ml/act_func/none.h"
#include "ml/act_func/relu.h"
#include "ml/act_func/tanh.h"
#include "ml/conv_layer/conv.h"
#include "ml/conv_layer/max_pool.h"
#include "ml/dense_layer/dense.h"
#include "ml/factory/factory.h"
#include "ml/factory/stub/factory.h"
#include "ml/flatten_layer/flatten.h"

namespace ml::factory
{
// -----------------------------------------------------------------------------
ActFuncPtr Factory::actFunc(const act_func::Type type) 
{ 
    // Create the corresponding activation function based on type.
    switch (type)
    {
        case act_func::Type::Relu:
            return std::make_unique<act_func::Relu>();
        case act_func::Type::Tanh:
            return std::make_unique<act_func::Tanh>();
        default:
            return std::make_unique<act_func::None>();
    }
}

// -----------------------------------------------------------------------------
ConvLayerPtr Factory::convLayer(const std::size_t inputSize, const std::size_t kernelSize, 
                             const act_func::Type actFunc) 
{
    ///! @todo Replace conv_layer::stub::Conv with conv_layer::Conv when implemented.
    return std::make_unique<conv_layer::stub::Conv>(inputSize, kernelSize, actFunc);
}

// -----------------------------------------------------------------------------
DenseLayerPtr Factory::denseLayer(const std::size_t inputSize, const std::size_t outputSize, 
                                  const act_func::Type actFunc)
{
    return std::make_unique<dense_layer::Dense>(inputSize, outputSize, actFunc);
}

// -----------------------------------------------------------------------------
FlattenLayerPtr Factory::flattenLayer(const std::size_t inputSize) 
{
    ///! @todo Replace flatten_layer::stub::Flatten with flatten_layer::Flatten when implemented.
    return std::make_unique<flatten_layer::stub::Flatten>(inputSize);
}

// -----------------------------------------------------------------------------
ConvLayerPtr Factory::maxPoolLayer(const std::size_t inputSize, const std::size_t poolSize)
{
    ///! @todo Replace conv_layer::stub::MaxPool with conv_layer::MaxPool when implemented.
    return std::make_unique<conv_layer::stub::MaxPool>(inputSize, poolSize);
}

// -----------------------------------------------------------------------------
FactoryPtr create(const bool stub)
{
    if (stub) { return std::make_unique<stub::Factory>(); }
    return std::make_unique<Factory>();
}
} // namespace ml::factory

File: python/ml/factory/factory.py
"""Machine learning factory."""
from ml.act_func import factory as act_func_factory
from ml.act_func.types import ActFuncType
from ml.act_func.interface import IActFunc
from ml.conv_layer.interface import IConvLayer
from ml.conv_layer.stub import ConvStub, MaxPoolStub
from ml.dense_layer.interface import IDenseLayer
from ml.dense_layer.dense import Dense
from ml.factory.interface import IFactory
from ml.factory.stub import FactoryStub
from ml.flatten_layer.interface import IFlatten
from ml.flatten_layer.stub import FlattenStub

class Factory(IFactory):
    """Machine learning factory."""

    def act_func(self, act_func_type: ActFuncType) -> IActFunc:
        """Create an activation function.

        Args:
            act_func_type: The type of activation function to create.

        Returns:
            The new activation function."""
        return act_func_factory.create(act_func_type)

    def conv_layer(self, input_size: int, kernel_size: int,
                   act_func_type: ActFuncType) -> IConvLayer:
        """Create a convolutional layer.
        
        Args:
            input_size: Input size. Must be greater than 0.
            kernel_size: Kernel size. Must be greater than 0 and smaller than the input size.
            act_func: Activation function to use.

        Returns:
            The new convolutional layer.
        """
         # Todo: Replace ConvStub with Conv when implemented.
        return ConvStub(input_size, kernel_size, act_func_type)

    def dense_layer(self, input_size: int, output_size: int,
                    act_func_type: ActFuncType) -> IDenseLayer:
        """Create a dense layer.
        
        Args:
            input_size: Input size. Must be greater than 0.
            output_size: Output size. Must be greater than 0.
            act_func_type: Activation function to use.

        Returns:
            The new dense layer.
        """
        return Dense(input_size, output_size, act_func_type)

    def flatten_layer(self, input_size: int) -> IFlatten:
        """Create a flatten layer.

        Args:
            input_size: Input size. Must be greater than 0.

        Returns:
            The new flatten layer.
        """
        # Todo: Replace FlattenStub with Flatten when implemented.
        return FlattenStub(input_size)

    def max_pool(self, input_size: int, pool_size: int) -> IConvLayer:
        """Create a max pooling layer.
        
        Args:
            input_size: Input size. Must be greater than 0.
            pool_size: Pool size. Must divide the input size.

        Returns:
            The new max pooling layer.
        """
        # Todo: Replace MaxPoolStub with MaxPool when implemented.
        return MaxPoolStub(input_size, pool_size)

def create(stub: bool = False) -> IFactory:
    """Create a factory.

    Args:
        stub: True to create a stub factory (default = false).

    Returns:
        The new factory.
    """
    return FactoryStub() if stub else Factory()
